File: BitLoc.js
const fs = require("fs");
const readline = require("readline");

const clear = () => process.stdout.write("\x1Bc");

const ASCII = c => [...c].map(ch =>
    [...Array(8)].map((_, i) => (ch.charCodeAt(0) >> (7 - i)) & 1)
).flat();

const ask = q => new Promise(res => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    rl.question(q, ans => { rl.close(); res(ans); });
});

async function run(program) {
    let mem = {}, ip = 0, buf = [];
    clear();

    while (ip >= 0 && ip < program.length) {
        let [bit, loc] = program[ip];

        mem[ip] = bit;
        buf.push(bit);

        if (buf.length === 8) {
            process.stdout.write(String.fromCharCode(parseInt(buf.join(""), 2)));
            buf = [];
        }

        if (mem[loc] === bit) {
            const inp = (await ask("")).trim();
            const bits = inp ? ASCII(inp) : Array(8).fill(0);

            mem[ip] = bits[0];
            for (let i = 1; i < bits.length; i++) mem[loc + i - 1] = bits[i];
        }

        ip = loc;
    }
}

async function repl() {
    while (true) {
        clear();
        console.log("BitLoc REPL");
        const file = (await ask("Enter File (or Enter to quit): ")).trim();
        if (!file) return console.log("Exiting REPL.");

        try {
            const program = fs.readFileSync(file, "utf8")
                .trim()
                .split("\n")
                .map(l => l.trim().split(/\s+/).map(Number));

            clear();
            await run(program);
        } catch (e) {
            console.log("Error:", e.message);
        }

        await ask("\nEnter to Continue REPL: ");
    }
}

repl();

File: BitLoc.py
import os
def clear():
    os.system('cls' if os.name == 'nt' else 'clear')
def ASCII(c):
    # converter
    val = ord(c)
    return [(val >> i) & 1 for i in range(7, -1, -1)]
def char_to_bits(ch):
    return ASCII(ch)
def run(program):
    memory = {}          # sparse tape
    ip = 0
    buffer = []          # collect 8 bits
    clear()
    prog_len = len(program)
    while 0 <= ip < prog_len:
        bit, loc = program[ip]
        # Write bit and maybe output a byte
        memory[ip] = bit
        buffer.append(str(bit))
        if len(buffer) == 8:
            print(chr(int("".join(buffer), 2)), end="")
            buffer.clear()
        # Conditional jump with input hijack
        if memory.get(loc) == bit:
            user_input = input().strip()
            if user_input:
                bits = [b for ch in user_input for b in char_to_bits(ch)]
            else:
                bits = [0] * 8
            # Overwrite A with first bit
            memory[ip] = bits[0]
            # Store remaining bits starting at loc
            for offset, b in enumerate(bits[1:]):
                memory[loc + offset] = b
        # Jump
        ip = loc
def repl():
    while True:
        clear()
        print("BitLoc REPL")
        file_name = input("Enter File (or press Enter to quit): ").strip()
        if not file_name:
            print("Exiting REPL.")
            break
        try:
            with open(file_name, 'r') as f:
                program = [
                    (int(a), int(b))
                    for line in f
                    for a, b in [line.strip().split()]
                ]
        except Exception as e:
            print(f"Error reading file: {e}")
            input("Press Enter to continue: ")
            continue
        clear()
        try:
            run(program)
        except Exception as e:
            print(f"\nError during execution: {e}")
        input("Enter to Continue REPL: ")
if __name__ == "__main__":
    repl()
